rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/MaximumBipartiteMatch.scala

```diff
@@ -8,9 +8,10 @@ import scala.Function.tupled
 import scala.collection.immutable.Stream
 
 object MaximumBipartiteMatch {
-  def apply[P, T, F[_]: Monad](matchFun: (P, T) => F[Boolean]): MaximumBipartiteMatch[P, T, F] = {
+  def apply[P, T, R, F[_]: Monad](
```

Comment:
> This function signature has changed but no tests are affected.

Because all of this is a refactoring + the last commit (memoize) is a slight optimization.

> Please write unit tests for the new functionality?

All new functionality has its unit tests added in this PR, because there's none :)

Reply:
I don't agree with you here. 

You say there is no new functionality but this component is not trivial and we should be able to write some unit tests which show how it works. I understand that there was no tests at this component level before but since we want to focus on "hardening the product" we should start adding such tests where they are missing.

Reply:
There's plenty of tests for what this component is being used for, namely the spatial matcher. If you manage to break a feature this code provides without causing a test failure in VarMatcherSpec, I'll gladly add a test for that. Notice though that memoization is not a feature - it's a performance optimization.

Reply:
In this case the tests are not about how the whole SpatialMatch works but about how this component i.e. MaximumBipartiteMatch works. It's not that much about validation but about documentation. 

I am a bit slow and I don't understand how it works. Can you explain how it works? If you can then do it with a few tests. It should be reasonably easy and will make the code easier to understand. The next person working in this area will thank you.

If we don't agree here then we will need another opinion from another reviewer on this subject.